warn/warn_macro.go

```diff
@@ -0,0 +1,334 @@
+// Warnings for using deprecated functions
+
+package warn
+
+import (
+	"fmt"
+
+	"github.com/bazelbuild/buildtools/build"
+)
+
+// Internal constant that represents the native module
+const nativeModule = "<native>"
+
+// acceptsNameArgument checks whether a function can accept a named argument called "name",
+// either directly or via **kwargs.
+func acceptsNameArgument(def *build.DefStmt) bool {
+	for _, param := range def.Params {
+		switch param := param.(type) {
+		case *build.Ident:
+			if param.Name == "name" {
+				return true
+			}
+		case *build.AssignExpr:
+			if ident, ok := param.LHS.(*build.Ident); ok && ident.Name == "name" {
+				return true
+			}
+		case *build.UnaryExpr:
+			if param.Op == "**" {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// fileData represents information about rules and functions extracted from a file
+type fileData struct {
+	rules     map[string]int
+	functions map[string]map[string]funCall
```

Comment:
Why are there nested maps?

Reply:
The outer map is for functions defined in the file, the inner is for each distinct function call from inside the function definition (if a function calls `foo` several times, it's not useful to store `foo` more than once).